lib/charms/prometheus_k8s/v0/prometheus_scrape.py

```diff
@@ -474,6 +476,106 @@ def _sanitize_scrape_configuration(job) -> dict:
+        """Format the topology information into a verbose string."""
+        return 'juju_model="{}", juju_model_uuid="{}", juju_application="{}"'.format(
+            self.model, self.model_uuid, self.application
+        )
+
+    def as_dict(self) -> dict:
+        """Format the topology information into a dict."""
+        return dataclasses.asdict(self)
+
+    def as_dict_long_form(self):
+        """Format the topology information into a dict with keys having 'juju_' as prefix."""
+        return {
+            "juju_model": self.model,
+            "juju_model_uuid": self.model_uuid,
+            "juju_application": self.application,
+        }
+
+    def render(self, template: str):
+        """Render a juju-topology template string with topology info."""
+        return template.replace("%%juju_topology%%", self.as_str_long_form())
+
+
+def load_alert_rules_from_dir(
+    dir_path: Union[str, Path], topology: JujuTopology, *, recursive: bool = False
+) -> List[dict]:
+    """Load alert rules from rule files.
+
+    All rules from files for the same directory are loaded into a single
+    group. The generated name of this group includes juju topology.
+    By default, only the top directory is scanned; for nested scanning, pass `recursive=True`.
+
+    Args:
+        dir_path: directory containing *.rule files (alert rules without groups).
+        topology: a `JujuTopology` instance.
+        recursive: flag indicating whether to scan for rule files recursively.
+
+    Returns:
+        a list of prometheus alert rule groups.
+    """
+    alerts = defaultdict(list)
+    for path in filter(Path.is_file, Path(dir_path).glob("**/*.rule" if recursive else "*.rule")):
+        relpath = os.path.relpath(os.path.dirname(path), dir_path)
+
+        # Generate group name:
+        #  - prefix, from the relative path of the rule file;
+        #  - name, from juju topology
+        group_name = (
+            f"{'' if relpath == '.' else relpath.replace(os.path.sep, '_') + '_'}"
+            f"{topology.as_str_short_form()}_alerts"
+        )
+
+        logger.debug("Reading alert rule from %s", path)
+        with path.open() as rule_file:
+            # Load a list of rules from file then add labels and filters
+            try:
+                rule = yaml.safe_load(rule_file)
+            except Exception as e:
+                logger.error("Failed to read alert rules from %s: %s", path.name, e)
+            else:
+                try:
+                    # add "juju_" topology labels
+                    rule.get("labels", {}).update(topology.as_dict_long_form())
+
+                    # insert juju topology filters into a prometheus alert rule
+                    rule["expr"] = topology.render(rule["expr"])
+                except KeyError:
+                    logger.error("Invalid alert rule %s: missing an 'expr' property.", path.name)
+                else:
+                    alerts[group_name].append(rule)
```

Comment:
The same here, maybe it's worth extracting this code into a custom function.

Besides, shouldn't we raise custom exceptions if `Exception` and `KeyError` is catched?

Reply:
> it's worth extracting this code into a custom function

one option is to pass the `alerts` "by-ref":
`read_rule_file(alerts, group_name, path, rule_file, topology)`

another option is an inner function:
`read_rule_file(group_name, path, rule_file)`

wdyt?

> raise custom exceptions

Per discussion with @balbirthomas, and in line with the original code, file reading should proceed even if one of them is buggy.